### DISCUSSION TITLE
Expiration date in the OAuth2 Module

### DISCUSSION THREAD

**Discussion by bjarkihall** [microsoft/ALAppExtensions#14325] [Q&A] [answered]
As seen on [AJ Kauffmann's blogpost on S2S in AL](https://www.kauffmann.nl/2021/08/04/service-to-service-authentication-in-business-central-18-3-how-to-use-in-al/#:~:text=The%20OAuth2%20codeunit%20just%20returns%20an%20access%20token%20without%20any%20information%20about%20the%20expiration.), it's not possible to get the expiration date via the OAuth2 module.

Is there a recommended way of getting the expiration date, because the only way I'm seeing around it, is by invoking the http call explicitly instead of using the system module, which kind of defeats the purpose of the module in that case, right?

Is the expiration date to be added to OAuth2Impl? It didn't look like Dotnet ALAzureAdCodeGrantFlow had any helper function to retrieve the expiration date from the previous requests, so I guess it isn't stored/exposed anywhere?

**Comment by tinfister** [comment, marked as answer]
I use AcquireTokenAndTokenCacheByAuthorizationCode to get token cache, then Base64Convert.FromBase64(TokenCache) to get nice json from cache and there you can find AccessToken with expiration time.
![image](https://user-images.githubusercontent.com/29481795/144567189-f13f08c8-e1bf-4b6e-99dd-610fcf9ca503.png)

**Comment by JesperSchulz** [comment]
Maybe it is worthwhile making the retrieval of the expiration time more convenient? If it doesn't create any security concerns, we'd be up for a PR :-)

**Reply by tinfister** [reply]
Will see what if time permits.

**Reply by bjarkihall** [reply]
@JesperSchulz I think the response is out of our control (and the AL team at Microsoft, see [this warning](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow#successful-response-1)) and we can't make a PR because this module is highly reliant on a DLL we don't have access to (and can't therefor make any contributions to).

We could make a PR that extracts the expiration date from the token (with Base64 Converter and Json types) but would it even be approved if it can easily break at some point?
